mqtt_as/interfaces/__init__.py

```diff
@@ -0,0 +1,92 @@
+from uerrno import EINPROGRESS, ETIMEDOUT
+import usocket
+import uasyncio as asyncio
+
+
+async def _g():
+    pass
+
+
+_type_coro = type(_g())
+
+
+# If a callback is passed, run it and return.
+# If a coro is passed initiate it and return.
+# coros are passed by name i.e. not using function call syntax.
+def launch(func, tup_args):
+    res = func(*tup_args)
+    if isinstance(res, _type_coro):
+        res = asyncio.create_task(res)
+    return res
+
+
+class BaseInterface:
+    def __init__(self, socket=None):
+        # Legitimate errors while waiting on a socket. See uasyncio __init__.py open_connection().
+        self.BUSY_ERRORS = [EINPROGRESS, ETIMEDOUT]
+        self.socket = socket or usocket  # support for custom socket implementations
+        self._subs = []
+        self._state = None
+
+    async def connect(self):
+        """Serve connect request. Triggers callbacks if state changes"""
+        if await self._connect():
+            self._change_state(True)
+            return True
+        return False
+
+    async def _connect(self):
+        """Hardware specific connect method"""
+        # return True  # if connection is successful, otherwise False
+        raise NotImplementedError()
+
+    async def disconnect(self):
+        """Serve disconnect request. Triggers callbacks if state changes"""
+        if await self._disconnect():
+            self._change_state(False)
+            return True
+        return False
+
+    async def _disconnect(self):
+        """Hardware specific disconnect method"""
+        # return True  # if disconnect is successful, otherwise False
+        raise NotImplementedError()
+
+    async def reconnect(self):
+        """Serve reconnect request"""
+        return await self._reconnect()
+
+    async def _reconnect(self):
+        """Hardware specific reconnect method"""
+        if await self._disconnect():
+            return await self._connect()
+        return False
+
+    def isconnected(self):
```

Comment:
my choice would be ``is_connected`` because I think it's much more readable

Reply:
while I don't disagree, the micropython network module uses isconnected() and so does mqtt_as so I think it makes more sense to go with that.